docs/udf.rst

```diff
@@ -0,0 +1,179 @@
+    public String substring(final String value, final int startIndex, final int endIndex)
+
+
+Example UDF class
+-----------------
+
+The class below creates a UDF named ``multiply``. As can be seen this UDF can be invoked in 3
+different ways:
+
+- with two int parameters returning a long result.
+- with two long parameters returning a long result.
+- with two double parameters returning a double result.
+
+.. code:: java
+
+    @UdfDescription(name = "multiply", description = "multiplies 2 numbers")
+    public class Multiply {
+
+      @Udf(description = "multiply two ints")
+      public long multiply(final int v1, final int v2) {
+        return v1 * v2;
+      }
+
+      @Udf(description = "multiply two longs")
+      public long multiply(final long v1, final long v2) {
+        return v1 * v2;
+      }
+
+      @Udf(description = "multiply two doubles")
+      public double multiply(final double v1, double v2) {
+        return v1 * v2;
+      }
+
+    }
+
+Supported Types
+---------------
+
+The types supported by UDFs are currently limited to:
+
+- int, Integer
+- boolean, Boolean
+- long, Long
+- double, Double
+- String
+- java.util.Map (must be a map of supported types)
+- java.util.List (must be a list of supported types)
+
+Note: Complex types are not currently supported
+
+
+=========
+Deploying
+=========
+
+To deploy your UDFs you need to create a jar containing all of the classes required by the UDFs.
+If you depend on third-party libraries then this should be an uber-jar containing those libraries.
+Once the jar is created you need to deploy it to each KSQL server instance. The jar should be copied
+to the ``ext`` directory that is part of the KSQL distribution.
+
+The jars in the ``ext`` directory are only scanned at start-up, so you will need to restart your
+KSQL server instances.
+
+
+=====
+Usage
+=====
+
+Once your UDFs are deployed you can call them in the same way you would invoke any of the KSQL
+builtin functions, for example, using the ``multiply`` example above:
+
+.. code:: sql
+
+    CREATE STREAM number_stream (int1 INT, int2 INT, long1 BIGINT, long2 BIGINT)
+      WITH (VALUE_FORMAT = 'JSON', KAFKA_TOPIC = 'numbers');
+
+    SELECT multiply(int1, int2), multiply(long1, long2) FROM number_stream;
+
+
+
+============
+Blacklisting
+============
+
+You can blacklist classes and packages such that they can't be used from a UDF. There is small
+blacklist that is found in the file ``resource-blacklist.txt`` that is in the ``ext`` directory.
+This file contains an entry per line, where each line is a class or package that should be blacklisted.
+The matching of the names is based on a regular expression, so if you have an entry, ``java.lang.Process``
+
+.. code:: txt
+
+    java.lang.Process
+
+This would match any paths that begin with java.lang.Process, i.e., java.lang.Process, java.lang.ProcessBuilder etc.
+
+Any blank lines or lines beginning with ``#`` are ignored. If the file is not present then all classes
```

Comment:
This is confusing - shouldn't it be that, if the file is not present, then all classes are whitelisted?

Reply:
If the file is empty all classes are whitlisted. If isn't present then they are all blacklisted. We could change it

Reply:
My expectation as a user would be:

- If the blacklist file is missing, nothing is blacklisted = all is whitelisted.
- If the blacklist file exists but is empty, nothing is blacklisted = all is whitelisted.
- If the blacklist file exists but contains (say) `*`, all/everything is blacklisted.

That being said, this raises the question of "out of the box" security for KSQL.  I read your PR as implying that, "typically you do want to blacklist some stuff to prevent people from wreaking havoc with UDFs", correct?  If yes, then perhaps:

1. We should implement the file logic as per my expectation above.
2. Include a default blacklisting file with a "good" list of blacklists.

Thoughts?

Questions:
- Alternatively, why not go with a whitelist approach, i.e. a whitelist file?  Everything is blacklisted unless explicitly whitelisted?
- What about transitive dependencies?  Would it be easier to configure "the file" if it is a whitelist vs. a blacklist file?

Reply:
@dguy : What about my questions above?

Reply:
@miguno, the first 2 points you make are ok. I think we should just scrap the blacklist everything option - in this case the UDFs will all fail. If you wan't to turn UDFs off then you can set the config to disable them.

I think the default blacklist, should actually be an empty one. People can decide for themselves what they do an don't want to block.

Anyway, that is not how it currently works, so the code would have to change.

Having a whitelist file would be way too much effort on the users behalf. They would be constantly chasing the items to put in it, i.e., anything that is loaded from a third party library etc. The whitelist would be come huge.

Having a blacklist for classes/packages that you definitely don't want seems like the simplest option at the moment. I don't see any point in further complicating it at this point.

Reply:
Sounds good.

> Anyway, that is not how it currently works, so the code would have to change.

So, what should we do? I agree with what you said:

> I think the default blacklist, should actually be an empty one. People can decide for themselves what they do an don't want to block.

Reply:
I'm going to update it, i just want the PR that makes this change to the code to go in first.
https://github.com/confluentinc/ksql/pull/1522

Reply:
Ack, LGTM!